pkg/agent/controller/egress/egress_controller.go

```diff
@@ -0,0 +1,674 @@
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (Egress name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncEgress(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Egress %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+// getOrAllocateMark gets or allocates a mark for a local Egress. Multiple Egresses can share same Egress IP.
+// The mark is per Egress IP, not per Egress.
+func (c *Controller) getOrAllocateMark(egressName, egressIP string) (uint32, error) {
+	c.egressIPStatesMutex.Lock()
+	defer c.egressIPStatesMutex.Unlock()
+
+	ipState, exists := c.egressIPStates[egressIP]
+	// Create an egressIPState if this is the first Egress using the IP.
+	if !exists {
+		ipState = &egressIPState{
+			egressIP:    net.ParseIP(egressIP),
+			egressNames: sets.NewString(egressName),
+		}
+		c.egressIPStates[egressIP] = ipState
+	} else if !ipState.egressNames.Has(egressName) {
+		ipState.egressNames.Insert(egressName)
+	}
+
+	// Ensure the Egress IP has a mark allocated.
+	var err error
+	if ipState.mark == 0 {
+		ipState.mark, err = c.idAllocator.allocate()
+		if err != nil {
+			return 0, fmt.Errorf("error allocating mark for IP %s: %v", egressIP, err)
+		}
+	}
+	// Ensure datapath is configured properly.
+	if !ipState.datapathInstalled {
+		if err := c.ofClient.InstallSNATMarkFlows(ipState.egressIP, ipState.mark); err != nil {
+			return 0, fmt.Errorf("error installing SNAT mark flows for IP %s: %v", ipState.egressIP, err)
+		}
+		if err := c.routeClient.AddSNATRule(ipState.egressIP, ipState.mark); err != nil {
+			return 0, fmt.Errorf("error installing SNAT rule for IP %s: %v", ipState.egressIP, err)
+		}
+		ipState.datapathInstalled = true
+	}
+	return ipState.mark, nil
+}
+
+func (c *Controller) releaseMark(egressName, egressIP string) error {
+	c.egressIPStatesMutex.Lock()
+	defer c.egressIPStatesMutex.Unlock()
+
+	ipState, exist := c.egressIPStates[egressIP]
+	// The Egress IP was not configured before, do nothing.
+	if !exist {
+		return nil
+	}
+	// Unlink the Egress from the EgressIP. If it's the last Egress referring to it, uninstall its datapath rules and
+	// release the mark.
+	ipState.egressNames.Delete(egressName)
+	if len(ipState.egressNames) > 0 {
+		return nil
+	}
+	if ipState.mark != 0 {
+		if err := c.ofClient.UninstallSNATMarkFlows(ipState.mark); err != nil {
```

Comment:
Same here - can we release lock before uninstall flows/rule?

Reply:
ditto